File: src/ramatlab/unsupervised/hpa.m
%{
    hpa - Aggregate rankings using the HPA algorithm and output results to a CSV file.
    
    This function imports the Python module `hpa` and calls its `hpa` function to perform
    rank aggregation based on input data provided in a CSV file. The output is also saved in a 
    specified CSV format.

    Reference:
    ----------
    - Fujita, S., Kobayashi, H., & Okumura, M. (2020). Unsupervised ensemble of ranking models for news comments using pseudo answers. In Advances in Information Retrieval: 42nd European Conference on IR Research, ECIR 2020, Lisbon, Portugal, April 14–17, 2020, Proceedings, Part II 42 (pp. 133-140). Springer International Publishing.

    Authors:
    --------
    Qi Deng

    Date:
    -----
    2024-10-13
%}

function hpa(input_file_path, output_file_path, input_type)
    % hpa - Calls the Python HPA algorithm for rank aggregation.
    %
    % This function imports the Python module `hpa` and calls its `hpa` function.
    % It processes the input CSV file containing voting data and outputs the aggregated rankings 
    % to the specified output CSV file.
    %
    % Parameters
    % ----------
    % input_file_path : str
    %     Path to the input CSV file containing voting data. The input should be in CSV format 
    %     with the following columns:
    %     - Query: Does not require consecutive integers starting from 1.
    %     - Voter Name: Allowed to be in string format.
    %     - Item Code: Allowed to be in string format.
    %     - Item Score/Item Rank: Represents the score/rank given by each voter. 
    %
    % output_file_path : str
    %     Path to the output CSV file where the aggregated rankings will be written.
    %
    % input_type : InputType
    %     The type of input data, which determines the interpretation of the fourth column 
    %     as either 'Item Rank' or 'Item Score'.
    
    hpamodule = py.importlib.import_module('hpa');
    hpamodule.hpa(input_file_path, output_file_path, input_type);
end

File: src/rapython/unsupervised/hpa.py
"""
HPA Algorithm

This implementation is based on the following reference:

Reference:
-----------
- Fujita, S., Kobayashi, H., & Okumura, M. (2020). Unsupervised ensemble of ranking models for news comments using pseudo answers. In Advances in Information Retrieval: 42nd European Conference on IR Research, ECIR 2020, Lisbon, Portugal, April 14–17, 2020, Proceedings, Part II 42 (pp. 133-140). Springer International Publishing.

Authors:
    Qi Deng
Date:
    2024-10-13
"""
import numpy as np

from src.rapython.common import InputType
from src.rapython.datatools import *

__all__ = ['hpa']


def hpa_func(sim, topk):
    # Get the number of gallery items and the number of rankers
    gallerynum = sim.shape[0]
    rankernum = sim.shape[1]

    # Calculate the average rank for each gallery item
    averagerank = np.sum(sim, axis=1)
    averagerank = averagerank / np.max(averagerank)  # Prevent division by zero

    # Obtain the ranking list by sorting in descending order
    ranklist = np.argsort(-sim, axis=0)

    # Initialize ndcg (Normalized Discounted Cumulative Gain)
    ndcg = np.zeros(rankernum)

    # Calculate ndcg values for each ranker
    for i in range(rankernum):
        for j in range(topk):
            # Update ndcg using the average rank and logarithmic scaling
            ndcg[i] += averagerank[ranklist[j, i]] * np.log(2) / np.log(i + 2)

    # Get the ranking indices of ndcg values in descending order
    ndcgrank = np.argsort(-ndcg)

    # Initialize finalrank to hold the cumulative scores
    finalrank = np.zeros(gallerynum)

    # Calculate the final ranking based on ndcg values and similarity scores
    for i in range(rankernum):
        finalrank += ndcg[ndcgrank[i]] * sim[:, ndcgrank[i]]

    # Sort the final ranking and transform the indices back to their original order
    finalrank = np.argsort(-finalrank)
    finalrank = np.argsort(finalrank)

    return finalrank


def hpa_agg(sim, topk):
    rankernum = sim.shape[0]
    querynum = sim.shape[1]
    item_num = sim.shape[2]

    result = np.zeros((querynum, item_num))

    for i in range(querynum):
        finalrank = hpa_func(sim[:, i, :].reshape(rankernum, item_num).T, topk)
        result[i, :] = finalrank.flatten()
    return result


def hpa(input_file_path, output_file_path, input_type, topk=10):
    """
    Process the input CSV file to aggregate rankings and write the results to an output CSV file.
    Parameters
    ----------
    input_file_path : str
        Path to the input CSV file.
        The input to the algorithm should be in CSV file format with the following columns:

        - Query: Does not require consecutive integers starting from 1.
        - Voter Name: Allowed to be in string format.
        - Item Code: Allowed to be in string format.
        - Item Score/Item Rank: Represents the score/rank given by each voter. It is recommended to choose the score format
    output_file_path : str
        Path to the output CSV file.
    input_type : InputType
        The type of input data. It determines the naming of the fourth column, which will either be 'Item Rank'
        or 'Item Score' based on this value.
    topk : int, optional
        The number of top-ranked items considered when calculating ranking metrics like ndcg (Normalized Discounted Cumulative Gain).
        Only the top `k` items from the ranker are included in the evaluation, where `k` is specified by this parameter.
        Defaults to 10.
    """
    input_type = InputType.check_input_type(input_type)
    df, unique_queries = csv_load(input_file_path, input_type)
    numpy_data, queries_mapping_dict = df_to_numpy(df, input_type)
    save_as_csv(output_file_path, hpa_agg(numpy_data, topk), queries_mapping_dict)
